### PR TITLE
Pie 1616/add payment recording in payment modal

### LINKED ISSUE #1613: I can click Record New Payment from my dashboard
## 🇺🇸 States this change applies to
- [X] ALL
- [ ] Nebraska
- [ ] Illinois

## 🧶 Acceptance Criteria

- [x] I can click "Record New" in the top right of my Dashboard.
- [x] The dropdown includes "Attendance Record" and "Payment." This replaces the current "Add Attendance" button.  
- [x] I can select "Payment," and the Record a Payment dialogue box will open 
- [x] The above acceptance criteria work in English and Spanish
- [ ] I have checked for accessibility (see details in PR template)

## 🥧 Implementation Details
Spanish translation: https://docs.google.com/spreadsheets/d/1d0RPc9PXvTCFpq8bfkZxgNL35MEpX5s8EvBwqex8564/edit#gid=413558466

Figma: https://www.figma.com/file/uCveVQQnYcw3MsenD1PNcG/P4P-2.0?node-id=5969%3A43540

### LINKED ISSUE #1616: I can record different payment amounts in the Record a Payment dialogue box
## 🇺🇸 States this change applies to
- [X] ALL
- [ ] Nebraska
- [ ] Illinois

## 🧶 Acceptance Criteria

- [ ] I can check the box to record a different revenue amount
- [ ] I can type in the "different amount" box only when I've checked the box. Otherwise, the input field is disabled. 
- [ ] I can only enter numbers in the input field
- [ ] The above acceptance criteria work in English and Spanish
- [ ] I have checked for accessibility (see details in PR template)

## 🥧 Implementation Details
Spanish translation: https://docs.google.com/spreadsheets/d/1d0RPc9PXvTCFpq8bfkZxgNL35MEpX5s8EvBwqex8564/edit#gid=413558466

Figma: https://www.figma.com/file/uCveVQQnYcw3MsenD1PNcG/P4P-2.0?node-id=5969%3A43540

### LINKED ISSUE #1617: I can hover over Calculated by Pie to see the tooltip (in the Record Payment dialogue box)
## 🇺🇸 States this change applies to
- [X] ALL
- [ ] Nebraska
- [ ] Illinois

## 🧶 Acceptance Criteria

- [ ] When I hover over "Calculated by Pie," the tooltip text appears (see Figma for copy) 
- [ ] The above acceptance criteria work in English and Spanish
- [ ] I have checked for accessibility (see details in PR template)

## 🥧 Implementation Details
Spanish translation: https://docs.google.com/spreadsheets/d/1d0RPc9PXvTCFpq8bfkZxgNL35MEpX5s8EvBwqex8564/edit#gid=413558466

Figma: https://www.figma.com/file/uCveVQQnYcw3MsenD1PNcG/P4P-2.0?node-id=5969%3A43540

### LINKED ISSUE #1618: I see the Record a Payment dialogue box with Earned Revenue for each child
## 🇺🇸 States this change applies to
- [X] ALL
- [ ] Nebraska
- [ ] Illinois

## 🧶 Acceptance Criteria

- [ ] Every child associated with my account appears in the dialogue box
- [ ] For each child, their earned revenue on the last day of the month appears
- [ ]  The dialogue box defaults to the previous month (e.g. if it is currently September, defaults to August). For now, this is a dropdown with only one option, until we add date filtering. 
- [ ] The dialogue box otherwise matches the mockup
- [ ] The above acceptance criteria work in English and Spanish
- [ ] I have checked for accessibility (see details in PR template)

## 🥧 Implementation Details
Spanish translation: https://docs.google.com/spreadsheets/d/1d0RPc9PXvTCFpq8bfkZxgNL35MEpX5s8EvBwqex8564/edit#gid=413558466

Figma: https://www.figma.com/file/uCveVQQnYcw3MsenD1PNcG/P4P-2.0?node-id=5969%3A43540

If possible, this earned revenue data should come from the already-implemented /case_list_for_dashboard endpoint, using last month as a filter date

### LINKED ISSUE #1619: I can save a payment total in my Record a Payment dialogue box
## 🇺🇸 States this change applies to
- [X] ALL
- [ ] Nebraska
- [ ] Illinois

## 🧶 Acceptance Criteria

- [ ] The number in the Save a Payment button dynamically adds up to the amounts listed above - EITHER the default amount OR the amount the user has entered for each row
- [ ] I can click the Save a Payment button to save that amount
- [ ] Relies on #1642 for the API call
- [ ] The above acceptance criteria work in English and Spanish
- [ ] I have checked for accessibility (see details in PR template)

## 🥧 Implementation Details
Spanish translation: https://docs.google.com/spreadsheets/d/1d0RPc9PXvTCFpq8bfkZxgNL35MEpX5s8EvBwqex8564/edit#gid=413558466

Figma: https://www.figma.com/file/uCveVQQnYcw3MsenD1PNcG/P4P-2.0?node-id=5969%3A43540

### LINKED ISSUE #1642: Each child's individual payment is saved by hitting the backend API
## ✨ What new feature do you want to see?

The API needs to accommodate saving a payment that is distributed across multiple children

## 🇺🇸 States this change applies to
- [X] ALL
- [ ] Nebraska
- [ ] Illinois

## 🧶 Acceptance Criteria

- [x] Blocked by #1644
- [ ] A `payments_controller` is created to accept these fields via a POST request
- [ ] A route is created for that controller method
- [ ] The appropriate request specs are created and pass

## 🥧 Implementation Details

add similar tests to:
```
context 'when sent with all required fields'
  it 'creates attendances and returns successful records' do
  context 'when the child has no active approval for that time period'
    it 'returns an error'

context 'when missing required fields on one record'
  it 'returns json errors'

context 'when missing required fields on all records'
  it 'returns json errors'

context "when adding an attendance for a child not in the user's care"
  it 'returns json errors'
```

### PR DESCRIPTION
**Pull request by davidcandeau**
## 💅🏼 What issue does this fix?

Closes #1613
Closes #1616
Closes #1617
Closes #1618
Closes #1619
Closes #1642

## 🍂 Before You Submit

* [x] Did you write & run tests and linters?
* [x] Does your PR contain any required translations?

### CODE REVIEW

File: client/src/Dashboard/DashboardTitle.js [hunk 1]
```diff
@@ -96,21 +103,67 @@ export default function DashboardTitle({ dates, userState, getDashboardData }) {
...
+      },
+      data: {
+        payments_batch: paymentsBatch
+      }
+    })
+
+    if (response.ok) {
+      setPaymentModalVisible(false)
+    } else {
+      // TODO: handle bad request
```
Comment (davidcandeau) [inline client/src/Dashboard/DashboardTitle.js:129]: @hnguyen0971 Can we handle this?

File: client/src/Dashboard/DashboardTitle.js [hunk 2]
```diff
@@ -96,21 +103,67 @@ export default function DashboardTitle({ dates, userState, getDashboardData }) {
...
-      id="paymentModal"
-      title={
-        <div className="eyebrow-large text-center">{t('recordAPayment')}</div>
-      }
+      className="payment-modal"
+      title={<div className="text-center h2-large">{t('recordAPayment')}</div>}
+      closeIcon={<CloseOutlined className="-btn-primary" />}
       visible={isPaymentModalVisible}
       onCancel={handlePaymentModalCancel}
+      //todo determine width. Maybe 50% of screen size
```
Comment (davidcandeau) [inline client/src/Dashboard/DashboardTitle.js:140]: I can fix this in #1623 as I will be playing around with sizing there

**Comment by davidcandeau** [comment]
I am sorry for mixing in a few issues into this PR. I got carried away while implementing the modal and did not realize there were separate issues for everything. 

The following issues should be tested in this PR:

#1616 
#1617 
#1618 
#1619

**Comment by katelovescode** [comment]
@rebeccakarasiktw I noticed some possibly unexpected behavior here.  When the box is deselected, the filler `$` remains, instead of reverting back to "Enter Amount", and also, if I've entered an amount, and then deselect the box, the number remains.  Do you have a preference on how these two scenarios should work?
clearbox
stickynumber

**Comment by katelovescode** [comment]
> #1616
Left a note for Rebecca because I'm unclear on expected behavior.

> #1617 
✅ 

> #1618

The unchecked box still gets calculated in the total, and gets sent and stored.

Screen Shot 2021-11-15 at 4 22 30 PM

Also @rebeccakarasiktw do we have desired UX behavior for when the child *has* no earned revenue for the last month?  Should they not be displayed?

> #1619
✅

**Comment by katelovescode** [comment]
@rebeccakarasiktw do you want the numbers to remain if the box is closed before submission, or to be cleared?
closereopen

@davidcandeau see the above gif - sometimes the button says $NaN when there are numeric values in the boxes

**Review by katelovescode — CHANGES_REQUESTED** [review]

**Comment by katelovescode** [comment]
I added some attendances for Manuel in my local for October, and his name and earned revenue aren't displaying, is this covered in another PR?
Screen Shot 2021-11-15 at 4 32 16 PM

File: client/src/Dashboard/DashboardTitle.js (hunk 2 above)
Comment (rebeccakarasiktw) [reply client/src/Dashboard/DashboardTitle.js:140]: Hi @davidcandeau @katelovescode yes, the width of the modal is 2/3 the width of the screen. We use a 24-column grid and the modal is 16 columns.

**Comment by katelovescode** [comment]
Review:
#1613 - ⚠️ One AC missing: "The dropdown includes "Attendance Record" and "Payment." This replaces the current "Add Attendance" button."
dropdown-english
dropdown-spanish

#1616 - ⚠️ Possible missing AC: "I can only enter numbers in the input field" - technically letters can be typed but if they are, the field is cleared when the user clicks or tabs away - cc @rebeccakarasiktw let me know if this is okay and I'll remove this comment
letters

#1617 - ✅ 

#1618 - ⚠️ Possible missing AC: "The dialogue box defaults to the previous month (e.g. if it is currently September, defaults to August). For now, this is a dropdown with only one option, until we add date filtering." Unless we decided not to make this a dropdown and I missed that convo
Screen Shot 2021-11-30 at 10 53 27 AM

#1619 - ✅ 
#1642 - ✅ 

Some weird behavior still on the input boxes:
The text box will clear if you uncheck the checkbox *unless* the amount is zero or empty
willnotclear0

I would not expect to click into the text box and be able to type within the phrase "Enter Amount" - i.e. this should be a placeholder but not typeable/erasable text.
enteramountistext

REALLY great progress here!

**Pull request merged**